src/Compilers/CSharp/Portable/Binder/SwitchBinder.cs

```diff
@@ -267,73 +253,62 @@ private BoundExpression ConvertCaseExpression(TypeSymbol switchGoverningType, CS
                     diagnostics.Add(ErrorCode.WRN_GotoCaseShouldConvert, node.Location, switchGoverningType);
                 }
 
-                convertedCaseExpression = sectionBinder.CreateConversion(caseExpression, conversion, switchGoverningType, diagnostics);
+                caseExpression = CreateConversion(caseExpression, conversion, switchGoverningType, diagnostics);
             }
 
-            if (switchGoverningType.IsNullableType()
-                && convertedCaseExpression.Kind == BoundKind.Conversion
-                // Null is a special case here because we want to compare null to the Nullable<T> itself, not to the underlying type.
-                && (convertedCaseExpression.ConstantValue == null || !convertedCaseExpression.ConstantValue.IsNull))
-            {
-                var operand = ((BoundConversion)convertedCaseExpression).Operand;
-
-                // We are not intested in the diagnostic that get created here
-                var diagnosticBag = DiagnosticBag.GetInstance();
-                constantValueOpt = sectionBinder.CreateConversion(operand, switchGoverningType.GetNullableUnderlyingType(), diagnosticBag).ConstantValue;
-                diagnosticBag.Free();
-            }
-            else
-            {
-                constantValueOpt = convertedCaseExpression.ConstantValue;
-            }
+            return ConvertPatternExpression(switchGoverningType, node, caseExpression, ref constantValueOpt, diagnostics);
+        }
 
-            return convertedCaseExpression;
+        private static readonly object s_nullKey = new object();
+        protected static object KeyForConstant(ConstantValue constantValue)
+        {
+            Debug.Assert(constantValue != (object)null);
+            return constantValue.IsNull ? s_nullKey : constantValue.Value;
         }
 
-        private List<SourceLabelSymbol> FindMatchingSwitchCaseLabels(ConstantValue constantValue, SyntaxNodeOrToken labelSyntax = default(SyntaxNodeOrToken))
+        protected SourceLabelSymbol FindMatchingSwitchCaseLabel(ConstantValue constantValue, CSharpSyntaxNode labelSyntax)
         {
             // SwitchLabelsMap: Dictionary for the switch case/default labels.
             // Case labels with a non-null constant value are indexed on their ConstantValue.
-            // Invalid case labels with null constant value are indexed on the labelName.
+            // Invalid case labels (with null constant value) are indexed on the label syntax.
 
             object key;
-            if (constantValue != null)
+            if (constantValue != (object)null && !constantValue.IsBad)
```

Comment:
I will move the cast to the left-hand-side throughout this source file.

---
In reply to: [70300895](https://github.com/dotnet/roslyn/pull/12416#discussion_r70300895) [](ancestors = 70300895)